File: src/daily/_202209/_20220902/daily.go
package _20220902

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func longestUnivaluePath(root *TreeNode) int {
	max, _ := path(root)
	return max
}

func path(root *TreeNode) (int, int) {
	if root == nil {
		return 0, 0
	}
	lm, lt := path(root.Left)
	rm, rt := path(root.Right)
	t := 0
	m := 0
	if root.Left != nil {
		if root.Val == root.Left.Val {
			t = 1 + lt
			m = 1 + lt
		}
	}
	if root.Right != nil {
		if root.Val == root.Right.Val {
			t = max2(1+rt, t)
			m += rt + 1
		}
	}
	return max3(lm, rm, m), t
}

func max2(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func max3(a, b, c int) int {
	return max2(max2(a, b), c)
}

File: src/daily/_202208/_20220820/Daily.kt
package daily._202208._20220820

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

class Solution {
    fun constructMaximumBinaryTree(nums: IntArray): TreeNode? {
        if (nums.isEmpty()) return null
        return recursiveConstruct(nums, 0, nums.size)
    }

    fun recursiveConstruct(nums: IntArray, start: Int, end: Int): TreeNode {
        var max = nums[start]
        var maxIndex = start
        for (i in start until end) {
            if (nums[i] > max) {
                max = nums[i]
                maxIndex = i
            }
        }
        val node = TreeNode(max)
        if (maxIndex > start) {
            node.left = recursiveConstruct(nums, start, maxIndex)
        }
        if (maxIndex < end - 1) {
            node.right = recursiveConstruct(nums, maxIndex + 1, end)
        }
        return node
    }
}
